### PR TITLE
Feature rbd merged

### PR DESCRIPTION
**Pull request by trickkiste**
Hi @alkisg,

I have finished my Ceph/RBD boot feature and everything works well.

As you proposed, I managed to do so by only using initramfs hooks and scripts. I kept them separate from the main ltsp initramfs magic. One reason being, that it will be easier to purge these file, if eventually rbd extensions will make it to upstream initramfs-tools.

I followed your proposed CMDLINE_METHOD path, but changed the naming to CMDLINE_BOOT_METHOD because it is more literal and therefore enables people to more quickly understand what is going on.

ltsp kernel should also work, if the RBD is mounted to the respective /srv/ltsp directory. It should then be treated like a chroot.

My best regards,
Markus

PS: Next step is to enable CephFS /home, which should be fairly easy in comparison.

### CODE REVIEW

File: debian/ltsp-rbd.initramfs-hook [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+    prereqs)
+        prereqs
+        exit 0
+        ;;
+esac
+
+. /usr/share/initramfs-tools/hook-functions
+
+modules="rbd aes cbc"
+manual_add_modules $modules
```
Comment (alkisg) [inline debian/ltsp-rbd.initramfs-hook]: If RBD is to be part of upstream LTSP, a single hook should be enough; there shouldn't be an additional hook file.
The modules"rbd aes cbc" should come from an INITRAMFS_MODULES ltsp.conf variable.

File: ltsp/common/info/55-info.sh [hunk 2]
```diff
@@ -32,4 +32,6 @@ info_main() {
     list_img_names -v
     printf "\nIMAGES:\n"
     list_img_names -i
+    printf "\nCMDLINE_BOOT_METHODS:\n"
+    list_img_names -m
```
Comment (alkisg) [inline ltsp/common/info/55-info.sh]: If CMDLINE_BOOT_METHODS is just an ltsp.conf variable, it's not significant enough to be shown as a separate section of `ltsp info`. People can just see it by reading ltsp.conf.

File: ltsp/common/ltsp/ltsp.conf [hunk 3]
```diff
@@ -14,6 +14,16 @@
...
+# Specify a CMDLINE_BOOT_METHOD for alternative boot methodes like RBD or iSCSI.
+# Here is an example RBD boot cmdline:
+# It follows this form: rbdroot=<mons>:<user>:<key>:<pool>:<image>[@<snapshot>]:[<partition>]:[<mountopts>]
+# Be aware that user is the username part only not "client.username" as often
+# referred to in Ceph. If the MON services listen on a non-standard port, port
+# shall be specified with ";" instead of standard ":" for obvious reasons -
+# e.g. host;7000
+# If you want to use the ${img} variable, you must use single quotes.
+# CMDLINE_BOOT_METHOD='boot=rbd ro rbdroot=10.101.0.4,10.101.0.5,10.101.0.6:user:secretkey==:rbd:${img}:1:'
+
```
Comment (alkisg) [inline ltsp/common/ltsp/ltsp.conf]: In the example ltsp.conf it would be best to include things that many people will need. I think RBD isn't too common, so I think it should be in the man page but not in the example ltsp.conf.

File: ltsp/server/ipxe/55-ipxe.sh [hunk 4]
```diff
@@ -43,6 +43,11 @@ ipxe_main() {
         items="${items:+"$items\n"}$(printf "item --key %d %-20s %s" "$((key%10))" "$img_name" "$title")"
         gotos=":$img_name\n$gotos"
     done
+    # FIXME: Alkis, it seems you are using the "r_" prefix here to prevent
+    # naming collisions. However, such collisions still would occure in
+    # /srv/tftp/ltsp, as you are not using a "roots" subdirectory to separate
+    # these namespaces. I have implemented the cmdline_boot_method items without
+    # the prefix in ipxe menu for that reason.
```
Comment (alkisg) [inline ltsp/server/ipxe/55-ipxe.sh]: If an image source exists in /srv/ltsp/chrootX and in /srv/ltsp/images/chrootX.img, then they should contain the same kernel/initrd, which is the one in /srv/tftp/ltsp/chrootX. It shouldn't be considered a collision.

File: ltsp/server/ipxe/55-ipxe.sh [hunk 5]
```diff
@@ -54,20 +59,36 @@ ipxe_main() {
...
 s|^/srv/ltsp|$BASE_DIR|g
 s/\(|| set menu-timeout \)5000/$(textif "$MENU_TIMEOUT" "\1$MENU_TIMEOUT" "&")/
 s|^:61:6c:6b:69:73:67\$|$(textif "$client_sections" "$client_sections" "&")|
-s|^#.*item.*\bimages\b.*|$(textif "$items$r_items" "$items\n$r_items" "&")|
-s|^:images\$|$(textif "$items" "$gotos" "&")|
+s|^#.*item.*\bimages\b.*|$(textif "$m_items$items$r_itmes" "$m_items\n$items\n$r_items" "&")|
 s|^:roots\$|$(textif "$r_items" "$r_gotos" "&")|
+s|^:images\$|$(textif "$items" "$gotos" "&")|
+s|^:cmdline_boot_method\$|$(textif "$m_items" "$m_gotos" "&")|
+s|\${CMDLINE_BOOT_METHOD}\$|$(textif "$CMDLINE_BOOT_METHOD" "$CMDLINE_BOOT_METHOD" "&")|
```
Comment (alkisg) [inline ltsp/server/ipxe/55-ipxe.sh]: So, I would only keep these, which would *replace* squashfs-over-nfs with squashfs-over-rbd. I wouldn't add more menu choices.

And other people could replace squashfs-over-nfs with squashfs-over-nbd, or squashfs-over-iscsi. We wouldn't want to generate all those menus just because we support multiple methods of booting.

File: ltsp/server/isc-dhcp-server/maas-dhcpd-snippet.conf [hunk 6]
```diff
@@ -0,0 +1,46 @@
+# This file is part of LTSP, https://ltsp.github.io
+# Copyright 2019 the LTSP team, see AUTHORS
+# SPDX-License-Identifier: GPL-3.0-or-later
+
+# MAAS isc-dhcp-server snippet for LTSP
+# in order to boot LTSP in a subnet add a global DHCP snippet and one to the
+# subnet, which you desire to enbale LTSP on.
```
Comment (alkisg) [inline ltsp/server/isc-dhcp-server/maas-dhcpd-snippet.conf]: Do we really want to ship a lot of isc-dhcp snippets based on what framework people are using?
If we can ship only one, it would make sense.
If more are needed, then I think the wiki is a more appropriate place to host them...

**Comment by alkisg** [comment]
Hi Markus,

I put some thoughts in some of the lines in the pull request.

I think there's enough information in your PR that I could use an integrate it to LTSP; I'm not asking you to do more work.

But at this point I would like to pause a bit and wait for **other user** feedback.
If you're the only one that needs it, then I would like to integrate only enough parts so that you can use RBD without having to modify LTSP code. So I would e.g. include the INITRAMFS_MODULES parameter, but not the RBD initramfs boot script.
If over the time, many people express interest, then it would surely make sense to have all of it completely automated.

File: debian/ltsp-rbd.initramfs-hook (hunk 1 above)
Comment (trickkiste) [reply debian/ltsp-rbd.initramfs-hook]: Hi @alkisg,

I am a fan of separation of concern. It keeps code clean, readable, understandable and more easy to maintain. I see no reason to merge everything into one file and no apparent benefit.

True, especially the hook file is pretty darn empty and it may seem obvious that it should be merged or INITRAMFS_MODULES should be used, but the logical conclusion would be, to merge the rbd script file as well. And that certainly does not make sense. So we should refrain from from doing so for the hook either, even it if appears to be so easy and practical.

The hook and the script are a unity, which I believe should not be broken up for just the benefit, that there is one file less. If somebody wants to research how this RBD feature actually works, scripts and hooks dirs are the most obvious places to search for the functionality, so there it should reside. To go looking in ltsp.conf is rather far off, if you approach this with existing initramfs knowledge but are new to LTSP. 

I'd rather follow the initramfs way to do things here than force people to spend time figuring out our new approach.

Also, as said, I believe code should invite to be reused. Like I reused this existing rbd initramfs code. And having these files well names and placed in hooks and scripts dirs facilitates that best.

File: ltsp/common/ltsp/ltsp.conf (hunk 3 above)
Comment (trickkiste) [reply ltsp/common/ltsp/ltsp.conf]: Ceph/RBD, for its apparent benefits, will become increasingly popular, I believe. Ceph/RBD is already the defacto standard as an Openstack storage backend and gaining much popularity with Kubernetes as well.

There is a reason, why it is part of the mainline kernel and RedHat bought the Inkspace, the company developing it.

However, I believe it would not hurt to add an iSCSI example as well.

File: debian/ltsp-rbd.initramfs-hook (hunk 1 above)
Comment (alkisg) [reply debian/ltsp-rbd.initramfs-hook]: Adding the modules increases the initramfs size.
So when other people start implementing hooks for iscsi, nbd (which will also need nbd-client), encrypted disks etc, you'll end up with a huge initramfs that people won't want to use.

This is the main reason that initramfs-tools doesn't include everything; it needs to keep some things optional.

Your code doesn't make RBD optional, it forces these modules to everyone's initramfs.

If you study the initramfs-tools and dracut code, you'll see that they do have provisions to make things optional, similar to ltsp.conf (e.g. /etc/initramfs-tools/modules).

**Comment by alkisg** [comment]
Regarding the initramfs script integration. Suppose a user runs:

```shell
apt install rbd-initramfs
```

And **this** command then adds the initramfs script and the modules etc. That action then makes it opt-in **and** keeps the code separate as you wanted. To achieve this, the ltsp source package (or even an unrelated source package) would build an rbd-initramfs binary package.

File: ltsp/server/ipxe/55-ipxe.sh (hunk 5 above)
Comment (trickkiste) [reply ltsp/server/ipxe/55-ipxe.sh]: @alkisg, for clarification, my Ceph/RBD feature does not include squashfs. There are no images stored on a given Rados Block Device. Instead the respective Rados Block Device is the actual (partitioned) boot device, which can also be used to spin up a qemu/kvm virtual machine for doing maintenance on the image.

Also the snapshotting feature built into Rados allows admins to very conveniently maintain backup versions of their boot block devices!

So, it is not squashfs-over-rbd it is direct rbd boot.

However, I agree, there is a lot of repetition in the current 55-ipxe.sh code and portions could be merged in to more general patterns.

I don't completely understand what you are suggesting here. That you would put an IF statement in there and skip the :images and :roots, if CMDLINE_BOOT_METHOD is set? 

I thought about that also before, but I came to the conclusion, that even if you had RBD feature in use, you might still want to be able to have one of the other boot methods in the menu in parallel - for testing something quickly.

One thought I had however was, that :cmdline_boot_method actually supports all the other methods as well. So this is the most generalised target and could replace the other two. They would then just set their own cmdline_boot_method before they jump to this target.

NFS, iSCSI, NBD, FibreChannel or something completely different (e.g. HTTP) could then all be supported through the same :cmdline_boot_method.

File: ltsp/server/isc-dhcp-server/maas-dhcpd-snippet.conf (hunk 6 above)
Comment (trickkiste) [reply ltsp/server/isc-dhcp-server/maas-dhcpd-snippet.conf]: @alkisg, I fully agree, we should not feel compelled to provide configs for any possible scenario.

However, I propose to include the full isc-dhcp-server example in the package and reference it in the man page for people to find more easily.

My version with the "cut marks" for MAAS snippets probably is better suited for the Wiki.

File: ltsp/server/ipxe/55-ipxe.sh (hunk 5 above)
Comment (alkisg) [reply ltsp/server/ipxe/55-ipxe.sh]: I can boot a VM disk file with LTSP like this:
mount flat-image.vmdk over NFS,
loop-mount it in /dev/loop8,
and then mount the first partition, /dev/loop8p1 to /root (the place where initramfs puts the root partition).

In an unrelated method, I can also call `ltsp image` to that file, and create a squashfs file, that is 2.5 times smaller and has less metadata, which then gives 2.5 times faster disk access.

And a third method is to tell LTSP to boot a disk and then get the squashfs image from inside it; this is used e.g. when booting a live-cd.iso with LTSP.

I do not know CephFS/RBD; help me understand. How would I boot a compressed squashfs file over CephFS, so that it's 2.5 times faster? Is CephFS the networking part, similar to NFS, so I could just use CephFS without RBD then?
Additionally, would I be able to partition and RBD device, and create a file system inside it, and put a squashfs file there, and ask LTSP to only boot that file?

**Comment by trickkiste** [comment]
> `apt install rbd-initramfs`

Seems like a good idea, in case initramfs-tools people don't want to include RBD feature directly in their package.

File: ltsp/server/ipxe/55-ipxe.sh (hunk 5 above)
Comment (trickkiste) [reply ltsp/server/ipxe/55-ipxe.sh]: BTW RBDs would also support compression. No need for squashfs.

**Comment by trickkiste** [comment]
> So let's wait a bit, and decide whether to do action 1 or action 2 based on user feedback.
Understandably, I am not all to happy about this.

RBD seems the be to logical choice for LTSP as a successor to all other non-redundant boot methods. It allows for a true "Cluster Setup", where each component is provided in a redundant manner with no-downtime fail-over.

Same is true for CephFS for /home. And its linear scalability makes it best suited for large school and university deployments.

Now, that I am up to date on the inner workings of this next gen LTSP version, I can surely manage to get my installations working without the code being fully included in upstream, but I would most certainly prefer, if it became part of mainline LTSP soon.

My best regards,
Markus

File: ltsp/common/info/55-info.sh (hunk 2 above)
Comment (trickkiste) [reply ltsp/common/info/55-info.sh]: Hi @alkisg,
the reasoning behind that is as follows:

Booting via CMDLINE_BOOT_METHOD is special, because in most cases (iSCSI, NBD, RBD), you do not create a squashfs but boot from the respective network block device directly. 

Therefore the theses need to be treated differently by certain tools and for that reason I had to include a specialised list_img_names() option in 55-images.sh.

Boot options coming from CMDLINE_BOOT_METHOD might also coexist with classic images and chroots in the ipxe menu. Therefore I would also list "images" (which they actually are not) separately in ltsp info.

Comment (alkisg) [reply ltsp/common/info/55-info.sh]: The design logic was as follows:

- chroots are the directory-based exports that are listed in iPXE
- images/* are the file-based exports that are listed in iPXE
- VMs are files that contain MBR, partitions, file systems etc; these don't show up in iPXE; they can be used as sources for `ltsp image`

What I propose now is:

- each directory-based export will only be listed once in iPXE, and be booted with NFS or whatever other network file system the sysadmin specifies
- each file-based export will only be listed once in iPXE, and be booted with NFS or NBD or iSCSI or whatever other block file system the sysadmin specifies

Sure, it would be possible to list a squashfs image 10 times in iPXE: bootable over NFS or NBD or iSCSI or AoE... but that would make the code too complicated, and the iPXE menu as well.
Experience from LSP5 says "keep it simple".

So I'd prefer to allow only one boot method for chroots, and one boot method for images. If the sysadmin needs more, he'd need to manually maintain his own ltsp.ipxe file.

Maybe we could name the ltsp.conf variables CHROOT_CMDLINE and IMAGE_CMDLINE, if it makes things more clear.

**Comment by alkisg** [comment]
Markus, how do RBD devices show up under /srv/ltsp?

They're block devices under /dev/rbd[0-9]*, and then you mount them in /srv/ltsp/cmdline_boot_method/* via /etc/fstab?
If so, can you paste an example fstab entry?

File: ltsp/server/ipxe/55-ipxe.sh (hunk 4 above)
Comment (trickkiste) [reply ltsp/server/ipxe/55-ipxe.sh]: :roots remains a little bit of a mystery to me for now.

Do I understand that correctly:
If one has a chroot you would create a squashfs image from it and boot from it by default but also make the actual chroot available to boot as NFS root without the squashfs_over_nfs detour? Possibly in rw mode so you can comfortably run maintenance tasks on it? 

Makes sense.

Would make sense for CMDLINE_BOOT_METHOD either!

But this would mean, a single CMDLINE_BOOT_METHOD config key for ltsp.conf is not enough to accommodate the parallel presentation of a "RO" boot method and a "RW" boot method.

On the other hand, if it were possible to define a CMDLINE_BOOT_METHOD_RO and a 
CMDLINE_BOOT_METHOD_RW or more generically CMDLINE_BOOT_METHOD_[0-9]* all current scenarios could be implemented using theses as well as RBD and anything coming in the future.

File: debian/ltsp-rbd.initramfs-hook (hunk 1 above)
Comment (trickkiste) [reply debian/ltsp-rbd.initramfs-hook]: @alkisg, 
I understand where you are coming from.

But, initrd.img as is in Ubuntu is already by far the largest thing we are loading to boot - 70M. The 4,6K of scripts/rbd don't really mean anything.

NFS is included by in initramfs-tools by default also. And no matter what you do in LTSP, I won't be able to make it optional if I am not using it, except if I lay hands on the file supplied by initramfs-tools.

However, RBD module loading and inclusion in the initrd could be made optional, if I'd include some magic, which looks at the CMDLINE_BOOT_METHOD in ltsp.conf and checks if boot=rbd is set.

Here is the exact accounting for the space used in initrd for rbd functionality

My current initrd including rbd:
elias@i7:~/src/debian$ ls -lh /srv/tftp/ltsp/ltsp-01/
insgesamt 78M
-rw-r--r-- 1 root root  70M Okt 23 02:15 initrd.img
-rw-r--r-- 1 root root 8,5M Okt 23 02:15 vmlinuz

--------------------------
sum = 714392 Bytes = 697K
--------------------------
-rw-r--r-- 1 root root 528913 Okt  1 00:23 /lib/modules/5.0.0-31-lowlatency/kernel/net/ceph/libceph.ko
-rw-r--r-- 1 root root 155673 Okt  1 00:23 /lib/modules/5.0.0-31-lowlatency/kernel/drivers/block/rbd.ko
-rw-r--r-- 1 root root  24321 Okt  1 00:23 /lib/modules/5.0.0-31-lowlatency/kernel/arch/x86/crypto/aes-x86_64.ko
-rw-r--r-- 1 root root   4640 Okt 22 22:55 /usr/share/initramfs-tools/scripts/rbd
-rw-r--r-- 1 root root    845 Okt 22 22:55 /usr/share/initramfs-tools/hooks/ltsp-rbd-next

File: ltsp/common/info/55-info.sh (hunk 2 above)
Comment (trickkiste) [reply ltsp/common/info/55-info.sh]: @alkisg, thanks for the clarification.

Please see my comment to topics below.

I believe there is a better way to generalise the setup and still allow chroots, images and even both in parallel. One thing missing though to make this work for everything:

BOOT_CMDLINE_[0-9]*
BOOT_DIR_[0-9]*

As one should be able to specify which boot method applies to the contents of which directory.

File: ltsp/server/ipxe/55-ipxe.sh (hunk 4 above)
Comment (trickkiste) [reply ltsp/server/ipxe/55-ipxe.sh]: As mentioned two topics above, a second ltsp.conf option would be needed to make this work for all scenarios:

BOOT_CMDLINE_[0-9]*     (renamed from CMDLINE_BOOT_METHOD)
BOOT_DIR_[0-9]*

File: ltsp/server/ipxe/55-ipxe.sh (hunk 5 above)
Comment (trickkiste) [reply ltsp/server/ipxe/55-ipxe.sh]: RBD - Rados Block Device, as the name indicates is a block device, which is stored in a Rados storage cluster in a redundant distributed manner and provided to the system via network.

It is rather comparable to NBD than NFS.
So you can partition this block device just like an HDD and format the partitions or the whole block device with any FS.

Snapshotting and Compression is built into RBD, so there is no need to go down the squashfs (read-only) detour.

You can have anybody boot from the device RO or from a snapshot and have one boot RW, so you can install stuff or administer users. Maybe we could even figure out a way to make the others ssh-auth to this particular host - just some thoughts anyways.

CephFS on the other hand is a fully POSIX compliant network file system, ergo more like NFS (which however is NOT fully POSIX compliant, which e.g. causes problems with Gnome dconf).

Every user could even have their own ceph access key, provided to them on ssh-auth and allowing them only to mount /home/user instead of all /home. That is the next part I will work on.

It is beautiful.

**Comment by trickkiste** [comment]
@alkisg, yes they show up as /dev/rbd[0-9]*/p[0-9]*.

And if you mount them, they show up where you want them.
Putting them in fstab is straight forward, as you just reference the device node and the FS options.
The actual RBD mapping is done prior to fstab through its own config file.

https://docs.ceph.com/docs/master/man/8/rbdmap/

Mapping on the LTSP client is actually done through /sys/bus/rbd/add_single_major.
https://www.kernel.org/doc/Documentation/ABI/testing/sysfs-bus-rbd

**Comment by trickkiste** [comment]
I like this conversation ;-)

**Comment by trickkiste** [comment]
Hi @alkisg,

I understand where you are coming from.

However, I want to propose a much cleaner more generic way to achieve the same thing. It is able to provide what you want but also allows for additional scenarios in a clean, generic and versatile way. I am sure you like it.

Let me just finish my draft and send you a pull request for it.

My best regards,
Markus

**Comment by stumbaumr** [comment]
Hi,
we are currently moving from a NetApp NFS based setup using LTSP to Ceph. Is this making it into LTSP soon or do we have to patch/fork ourselves?

Best regards
Rainer

**Comment by alkisg** [comment]
Hi Rainer,

this isn't going to be merged soon in its current form. I plan to experiment with Ceph and RDB next year; I might add proper support for it to LTSP then. In the meantime, I'm willing to accept small patches that make LTSP "not stand in the way" of someone that wants to use it with Ceph/RDB. Specifically, the following issues are to be addressed:

- A custom entry in ltsp.ipxe can already be provided if one uses a "localized template" override, /usr/share/ltsp/server/ipxe/ltsp-en.ipxe.
- Adding modules to initramfs can already be done as usual with initramfs-tools.
- Writing an initrd-mount hook that implements Ceph/RDB root mounting for initramfs-tools can be implemented outside of the LTSP codebase.

So, if after these ^, some existing LTSP scripts need patching at all, I wouldn't mind accepting such patches.

**Comment by trickkiste** [comment]
Hi Rainer, Hi Alkis,

I want to provide you with some of my conclusions from running LTSP on
Ceph/RDB and home directories on CephFS for a year now.

This is my current setup:

* LTSP boots from Ceph/RBD (Rados Block Devices == Network Block Devices,
which can hold partitions formated in all available FS types).
* Data is stored on this RBD in squashfs (for speed and size).
* I therefore create a squashfs image, which I then dd to the RBD.
* The image is created from a virtual machine (which also happens to live
on a Ceph/RBD, but that is not important).
* User accounts are currently stored on the VM host machine.
* I mount the VMs partition and then create the image.
* I generate the ltsp initrd image holding the user account and group info.
* Home lives on CephFS.
* The necessary Ceph/RBD auth keys to boot are provided via ipxe boot
options.
* CephFS auth info is on the boot image.

This setup works great, but updates are a bit cumbersome.

I will next try to boot everything from CephFS instead!

* I will put the whole root filesystem into CephFS and set read-only
permissions for the clients.
* Only /home will be writeable.
* It would also be possible, to restrict access to /home/username on
username basis and provide the auth keys on login.
* I will have a VM, which accesses the root-fs in write mode and therefore
I can install new software on the fly.
* The clients will boot of this CephFS in read-only mode. The can not
change anything, but changes are reflected in real-time. So I can install
new software and these programs are available on the already running
clients right away ... some Gnome menu reload mechanism might be needed.
* Also changes to passwd, shadow, group, gshadow would be reflected
immediately.

Depending on network read speed, things should run smoothly.

Alternatively, I could have this root-fs on CephFS, but generate squashfs
images from it.
Clients first mount the CephFS dir with the images in them, then boot those
images.

However, I like the first variant better, as updates are just less work
then.

Please provide me with your thoughts on this plan.

Alkis, I am especially interested in, whether LTSP can deal with this setup
from an overlayfs perspective, as the CephFS root FS will also hold /var
... things the clients pseudo write to via overlayfs currently.

Cheers,
Markus




Currently, I am booting LTSP from RBD (Rados Block Devices == Network Block
Devices, which can hold partitions formated in all available FS types)

>

**Comment by alkisg** [comment]
Hi Markus,

thank you for sharing your ideas; I can see some with the "trying to avoid squashfs" goal:

1. As you said, I'm not sure if CephFS will work as a lower overlayfs file system. You'd need to test or ask the kernel developers about that.
2. More importantly, I don't think overlayfs would be stable when the underlying file system was writable, instead of being read only, and just the tmpfs on top being writable.
3. And most importantly, I don't think the operating system itself would be stable, if "another operating system", the one in the server, was modifying system files, the ones in /usr or /var, without cooperating or notifying the operating system in the client. It would cause "corrupted operating system state", similar to "corrupted filesystems".

It's like asking operating systems to behave as databases do, which allow concurrent access from many sources, because they're specifically designed for that. But this isn't the case, no operating systems are designed that way.

What you **can** do for "3", is use snapshots, either in block devices or in filesystems that support them. The LTSP clients would use the previous snapshot, while you'd maintain the "trunk" of the boot device. When you'd be done, you'd create a newer snapshot, and LTSP clients would boot into that when they'd reboot.

To sum up, I do not think the ability to "apt install a program on the server, and have it show up in all fat clients without rebooting", is not possible without risking corruption. It's only possible for thin clients, since there's only one operating system in that case, the one on the server.

**Comment by alkisg** [comment]
Closing this one as documented in https://github.com/ltsp/ltsp/issues/37#issuecomment-1082695696. Thanks!
